### PR TITLE
Feat/fetch postby user

### LINKED ISSUE #1589: Feature Request: Implement GraphQL Query to Fetch Posts by UserID for Profile Page
**Is your feature request related to a problem? Please describe.**
Currently, in Talawa-Admin, we can fetch and display posts based on the organization ID. However, for the user profile section in Talawa-Mobile, we need the functionality to display posts created by individual users. The existing system does not support fetching posts filtered by userID.

**Describe the solution you'd like**
Implement a new GraphQL query named postsByUser (or any suitable name for the query), which takes a userID as an input and returns all posts created by that user. This will allow the user profile section in Talawa-mobile to display user-specific posts, enhancing the profile information with their activity.

**Describe alternatives you've considered**

Modifying Existing Query: Adjusting the current postsByOrganization query to also filter by userID. This could make the query overly complex and less efficient for its original purpose.

**Approach to be followed**
1.Update the GraphQL schema to include the postsByUser query.
2.Implement the resolver function for postsByUser, ensuring it queries the database for posts matching the provided userID.

**Additional context**
This feature is essential for a more personalized user experience in the profile section.

**Potential internship candidates**
Please read this if you are planning to apply for a Palisadoes Foundation internship https://github.com/PalisadoesFoundation/talawa/issues/359

### PR DESCRIPTION
**Pull request by AVtheking**
**What kind of change does this PR introduce?**

Implemented the feature of getting post by user

**Issue Number:**

Fixes #1589 

**Did you add tests for your changes?**

Yes

### CODE REVIEW

**Comment by palisadoes** [comment]
Please fix the failing tests

**Comment by AVtheking** [comment]
I have fixed the failing test case please review it

File: src/app.ts
Comment (palisadoes) [inline src/app.ts:12]: Remove and fix the root cause of the linting error

File: src/index.ts
Comment (palisadoes) [inline src/index.ts:44]: Remove and fix the root cause of the linting error

File: src/index.ts
Comment (palisadoes) [inline src/index.ts:57]: Remove and fix the root cause of the linting error

**Review by palisadoes — CHANGES_REQUESTED** [review]
See comments.

File: src/index.ts
Comment (AVtheking) [reply src/index.ts:57]: what would be the type of error here it is showing error for every possilbe type I applied

Comment (palisadoes) [reply src/index.ts:57]: https://typescript-eslint.io/rules/no-explicit-any/

Try using the playground here where they give suggestions.

**Comment by AVtheking** [comment]
I have fixed the linting error please review the changes

File: tests/helpers/posts.ts
Comment (palisadoes) [inline tests/helpers/posts.ts:9]: Remove. We need properly linted code.

File: tests/helpers/posts.ts
Comment (palisadoes) [inline tests/helpers/posts.ts:14]: Remove. We need properly linted code.

File: tests/helpers/userAndOrg.ts
Comment (palisadoes) [inline tests/helpers/userAndOrg.ts:7]: Remove. We need properly linted code.

File: tests/helpers/userAndOrg.ts
Comment (palisadoes) [inline tests/helpers/userAndOrg.ts:11]: Remove. We need properly linted code.

File: tests/resolvers/Organization/creator.spec.ts
Comment (palisadoes) [inline tests/resolvers/Organization/creator.spec.ts:73]: Remove. We need properly linted code.

**Review by palisadoes — CHANGES_REQUESTED** [review]
See comments

File: tests/helpers/userAndOrg.ts
Comment (AVtheking) [reply tests/helpers/userAndOrg.ts:11]: If I remove it from here then I have to edit all test files because each file have used 'any' type which would be out of scope of this pr shall I proceed to change all the 'any' type ??

Comment (AVtheking) [reply tests/helpers/userAndOrg.ts:11]: Or @palisadoes  sir I would solve this another pr ?

Comment (palisadoes) [reply tests/helpers/userAndOrg.ts:11]: You would only have to edit the test files submitted in this PR correct? That is in scope of this PR.
We prefer not to merge code without proper tests and linting that is why we have automation to verify compliance with our goals

Comment (AVtheking) [reply tests/helpers/userAndOrg.ts:11]: But sir I have to as in most of the test files there is use of type of any I have to edit all those files, if not then it shows linting error

Comment (palisadoes) [reply tests/helpers/userAndOrg.ts:11]: 1. We have created a linter that only lints on committed files.
2. This was merged about a week ago
3. Please merge your code with the latest upstream to see whether it is in effect for you

Comment (AVtheking) [reply tests/helpers/userAndOrg.ts:11]: Yes sir as their are many files which uses testUser and testOrganization therefore if I change type any of testUser and testOrganization I have to fix type in every file which uses these two and that will be a lot of files shall I do that ?

Comment (palisadoes) [reply tests/helpers/userAndOrg.ts:11]: 1. I see, it's hundreds of files. 
2. We'll need to fix this some other way

**Review by palisadoes — APPROVED** [review]

**Review by palisadoes — APPROVED** [review]

**Comment by palisadoes** [comment]
@JamaicanFriedChicken
@sumitra19jha
@xoldyckk

Please review. This looks ready

File: src/typeDefs/types.ts
Comment (xoldyckk) [inline src/typeDefs/types.ts:378]: `posts(after: String, before: String, first: Int, last: Int): PostsConnection`

you have to make use of a graphql connection, you can't just return a full blown list of objects all at once, I had mentioned this in the issue

you need to care about naming things correctly, this is standard practice

**Comment by xoldyckk** [comment]
@palisadoes the linting issues can only be fixed in an ad hoc way as long as mongoose is being used, mongoose doesn't work well with typescript, I have mentioned this in the past

the only known ORMs for mongodb that are designed to work well with typescript are `micro-orm` and `prisma`, other than that you have to hand-roll custom typescript types for your mongodb database objects which is too much work

File: src/typeDefs/types.ts
Comment (xoldyckk) [reply src/typeDefs/types.ts:378]: @AVtheking yes, cursor based pagination as is the standard for graphql

Comment (AVtheking) [reply src/typeDefs/types.ts:378]: Ok will learn and implement it

**Comment by AVtheking** [comment]
@xoldyckk  sir I have implemented the cursor based pagination please review

**Comment by AVtheking** [comment]
Please review and suggest changes if any

File: src/resolvers/User/post.ts
Comment (xoldyckk) [inline src/resolvers/User/post.ts:10]: stop making use of `as` keyword, it's an escape hatch to circumvent type safety, it should be used sparingly

look at this https://github.com/PalisadoesFoundation/talawa-api/blob/main/src/utilities/graphqlConnectionFactory.ts

the logic is a bit different but the implementation is correct, create a generic connection argument validation function which validates that:-
1. either `first` or `last` are provided
2. when `first` is provided, `last` isn't provided
3. either `first` and `after` or `last` and `before` are provided
4. when `first` and `after` are provided, `last` and `before` aren't provided

**Comment by AVtheking** [comment]
I have made all the requested changes ,please review them

File: src/utilities/validateConnectionArgs.ts
Comment (xoldyckk) [inline src/utilities/validateConnectionArgs.ts:1]: This is hard to make sense of. I'll just give you the code logic, here:-
```typescript
import { GraphQLError } from "graphql";

export function isNotNullish<T>(value: T | undefined | null): value is T {
  return value !== undefined && value !== null;
}

export type RelayConnectionArguments = {
  after?: string | null;
  before?: string | null;
  first?: number | null;
  last?: number | null;
};

export type ParsedRelayConnectionArguments = {
  cursor: string | null;
  limit: number;
  direction: "FORWARD" | "BACKWARD";
};

// some global or database record based hard limit for the number of records
// that can be returned in one connection query for the corresponding record
// replace this which an environment variable preferably
const env = {
  LIMIT: 0,
};

export function parseConnectionArguments(
  args: RelayConnectionArguments,
): ParsedRelayConnectionArguments {
  const paginationArgs: ParsedRelayConnectionArguments = {
    cursor: null,
    direction: "FORWARD",
    limit: 0,
  };

  if (isNotNullish(args.first)) {
    if (isNotNullish(args.last)) {
      throw new GraphQLError(
        "Argument first cannot be provided with argument last.",
      );
    }
    if (isNotNullish(args.before)) {
      throw new GraphQLError(
        "Argument before cannot be provided with argument first.",
      );
    }
    if (args.first > env.LIMIT) {
      throw new GraphQLError(
        `Argument first cannot not be greater than ${env.LIMIT}.`,
      );
    }
    paginationArgs.direction = "FORWARD";
    paginationArgs.limit = args.first;
    if (isNotNullish(args.after)) {
      paginationArgs.cursor = args.after;
    }
  } else if (isNotNullish(args.last)) {
    if (isNotNullish(args.after)) {
      throw new GraphQLError(
        "Argument after cannot be provided with argument last.",
      );
    }
    if (args.last > env.LIMIT) {
      throw new GraphQLError(
        `Argument last cannot be greater than ${env.LIMIT}.`,
      );
    }
    paginationArgs.direction = "BACKWARD";
    paginationArgs.limit = args.last;
    if (isNotNullish(args.before)) {
      paginationArgs.cursor = args.before;
    }
  } else {
    throw new GraphQLError(
      `Either argument first or argument last must be provided.`,
    );
  }

  return paginationArgs;
}

```

Comment (AVtheking) [reply src/utilities/validateConnectionArgs.ts:1]: Respected @xoldyckk sir Thank you for providing me code logic ,I want to ask that do I need to use this validation in my query code as I haven't used "Forward"
and "Backward" in my code logic ,do I need to change my code logic in the quey or just paste this code in validation file for future use ?

Comment (xoldyckk) [reply src/utilities/validateConnectionArgs.ts:1]: read this:- https://relay.dev/graphql/connections.htm#sec-Pagination-algorithm

take example of messages in a slack channel

on slack when loading older messages scrolling upwards, that is backward pagination

on slack when loading newer messages scrolling downwards, that is forward pagination((though in chat apps this is achieved through ws/sse and not the client manually fetching a list of latest messages)

though this forward/backward naming wouldn't make sense if you provided the slack clients ability to sort the messages in ascending/descending order

Comment (AVtheking) [reply src/utilities/validateConnectionArgs.ts:1]: Thank you sir for providing me resources , although I have understood the logic of forward and backward pagination I just wanted to ask do I need to change my code logic according to the validation logic which you have send me?

Comment (AVtheking) [reply src/utilities/validateConnectionArgs.ts:1]: Respected @xoldyckk  sir what would be the value of LIMIT variable ?

**Comment by palisadoes** [comment]
Please fix the failing tests. Our test code coverage system will fail if any of these conditions occur:

1) The overall code coverage drops below the target threshold of the repository
2) Any file in the pull request has code coverage levels below the repository threshold

We do this to ensure the overall reliability of the code base is maintained. This is a prerequisite for getting your PR approved. Assigned reviewers regularly review the PR queue and tend to focus on PRs that are passing.

File: src/utilities/validateConnectionArgs.ts
Comment (AVtheking) [reply src/utilities/validateConnectionArgs.ts:24]: Yes I have just added this as temp, was waiting for your reply

**Comment by AVtheking** [comment]
I have made the requested changes

**Comment by AVtheking** [comment]
Please review it and suggest changes if there is any

File: src/utilities/validateConnectionArgs.ts
Comment (xoldyckk) [inline src/utilities/validateConnectionArgs.ts:1]: rename the file to `parseRelayConnectionArguments.ts`

File: src/typeDefs/types.ts
Comment (xoldyckk) [inline src/typeDefs/types.ts:396]: `first` and `last` are the amount of records that client wants to fetch in one round trip to the server corresponding to their pagination direction

the `limit` that I asked you to be passable to `parseRelayConnectionArguments` function was to do with the allowed maximum records to be fetched in one round trip from the server, the point was to give each connection resolver the **granularity** in how many records they will return in one request

this `limit` isn't to be added to the graphql schema

File: src/utilities/validateConnectionArgs.ts
Comment (xoldyckk) [inline src/utilities/validateConnectionArgs.ts:20]: `limit` shouldn't be a part of the relay connection arguments, it is to be provided seperately

File: src/utilities/validateConnectionArgs.ts
Comment (xoldyckk) [reply src/utilities/validateConnectionArgs.ts:39]: `(args: RelayConnectionArguments, maximumLimit: number)`

File: src/utilities/validateConnectionArgs.ts
Comment (xoldyckk) [inline src/utilities/validateConnectionArgs.ts:58]: it's better to not error out here, make it this:-
```typescript
paginationArgs.direction = "FORWARD"
if(args.first > maximumLimit) {
  paginationArgs.limit = maximumLimit
} else {
  paginationArgs.limit = args.first
}
```

File: src/utilities/validateConnectionArgs.ts
Comment (xoldyckk) [inline src/utilities/validateConnectionArgs.ts:74]:
```typescript
paginationArgs.direction = "BACKWARD"
if(args.last > maximumLimit) {
  paginationArgs.limit = maximumLimit
} else {
  paginationArgs.limit = args.last
}
```

File: src/resolvers/User/posts.ts
Comment (xoldyckk) [inline src/resolvers/User/posts.ts:29]: make use of this function instead of manually creating the connection object https://github.com/PalisadoesFoundation/talawa-api/blob/8a0a2e447c2f86689302a5edf198aea2bcc84821/src/utilities/graphqlConnectionFactory.ts#L34

File: src/typeDefs/types.ts
Comment (AVtheking) [reply src/typeDefs/types.ts:396]: Ok sir will correct it

**Comment by AVtheking** [comment]
I have made changes

File: src/resolvers/User/posts.ts
Comment (xoldyckk) [inline src/resolvers/User/posts.ts:47]: the `$lt` and `$gt` filters are only to be used when cursor is not null

File: src/resolvers/User/posts.ts
Comment (xoldyckk) [inline src/resolvers/User/posts.ts:34]: if the cursor is null you have to ignore this check

File: src/resolvers/User/posts.ts
Comment (xoldyckk) [inline src/resolvers/User/posts.ts:35]: the point was to create the graphql connection at the beginning of resolver, and mutate accordingly it as the logic traverses down
```typescript
const postsConnection = graphqlConnectionFactory<InterfacePost>();
```
if the cursor is null you have to ignore this check, also no point seperating it out in a seperate function
```typescript
if(paginationArgs.cursor !== null) {
  const post = await Post.findById(cursor);
  if (!post) {
    return postsConnection
  }
}
```

**Comment by palisadoes** [comment]
Please fix the conflicting file

**Comment by AVtheking** [comment]
Respected @xoldyckk  sir do you want that I instead of typing manually should use the functions from the graphqlConnectonFactory.ts file ? if yes then please tell me which type should I specify in the generateConnectionObject funciton 
for t2 as I think t1 and t2 would be same for me .

**Comment by AVtheking** [comment]
which conflicting files are you talking about sir ? It is not showing me any conflicts

**Comment by AVtheking** [comment]
Please review it sir

**Comment by AVtheking** [comment]
@xoldyckk  Respected sir please review it ,it already for a long time this pr is lying dormant ,if there is any changes required I am ready to solve it.

**Comment by AVtheking** [comment]
I don't know why is this giving formatting error even though I am passing formatted code

**Comment by AVtheking** [comment]
on merging with the latest upstream it is showing already up to date,means my branch is already up to date

**Comment by palisadoes** [comment]
Did you run `prettier` against the files listed in the root y?

**Comment by AVtheking** [comment]
Yes sir

**Comment by palisadoes** [comment]
Ask the `talawa-api` slack channel. It's not happening to any other PRs

**Review by sumitra19jha — COMMENTED** [review]
Please don't do unnecessary changes. Because of your indentation settings in vs code, lot of changes are appearing. Make sure we only see the main code.

**Comment by AVtheking** [comment]
sorry sir for the inconvience but these changes in the file occured after merging and resolving conflicts

**Comment by palisadoes** [comment]
*NOTE* Read very carefully

1. We just merged this PR which upgraded the prettier package. 
    1. https://github.com/PalisadoesFoundation/talawa-api/pull/1840
1. It reformatted over 400 files. This will put your PR at risk of extensive merge conflicts.    
1. I suggest that before you merge your code to the latest upstream, or manually fix these merge conflicts that you:
    1. upgrade your `prettier` in your local branch to the same version
    1. run `prettier`
    
This will help to reduce the number of future merge conflicts for your PR.

**Comment by AVtheking** [comment]
Thank you sir it helps in reducing confilcts

**Comment by AVtheking** [comment]
> Please don't do unnecessary changes. Because of your indentation settings in vs code, lot of changes are appearing. Make sure we only see the main code.

Now I have removed those files which have unnecessary changes please review now

**Comment by palisadoes** [comment]
Please get the code coverage for the patch higher. It is significantly below the target

**Comment by AVtheking** [comment]
Sir  I have added tests for those missing lines but still it is not showing them covered

**Comment by palisadoes** [comment]
Is this ready for review?

**Comment by AVtheking** [comment]
> Is this ready for review?
Yes sir

**Comment by AVtheking** [comment]
I have made changes according to #1883 PR please review it now and tell me if there is any error

**Comment by AVtheking** [comment]
Also sir @xoldyckk  I would love to know the work of parseCursor function I didn't understand that function fully

**Comment by xoldyckk** [comment]
@AVtheking 

Each connection resolver may have different requirements for encoding the cursor into a string to be used inside the connection edge, this would be done inside the `createCursor` callback function. Consequently, those same connection resolvers will have different requirements for parsing and decoding the cursor accepted, this would be done inside the `parseCursor` callback function. All cursors could be ids but all ids are not cursors.

Another advantage is having each connection resolver return custom error messages inside their `parseCursor` callback function. The point is to parse all arguments received by a connection resolver and create a flatmap of errors that occurred pertaining to those arguments. So, instead of returning a single error related to a single argument to the client, we return all errors related to all arguments in one go. This reduces the amount of requests it would take for the server to communicate the errors to the client.

**Comment by AVtheking** [comment]
@xoldyckk  thank you sir for the explanation

**Comment by AVtheking** [comment]
this PR is ready for the reivew

File: src/typeDefs/queries.ts
Comment (xoldyckk) [inline src/typeDefs/queries.ts:110]: this should be exposed as `Organization.posts`

File: src/resolvers/User/posts.ts
Comment (xoldyckk) [inline src/resolvers/User/posts.ts:20]: Include the information about connection and pagination in tsdoc comments

File: tests/resolvers/User/post.spec.ts
Comment (xoldyckk) [inline tests/resolvers/User/post.spec.ts:75]: Remove the console.log statements, not necessary

File: tests/resolvers/User/post.spec.ts
Comment (xoldyckk) [inline tests/resolvers/User/post.spec.ts:97]: copy paste the assertion format like I've done for `UserTag/childTags.spec.ts`, it would be cleaner

**Comment by xoldyckk** [comment]
`package-lock.json` is modified without `package.json` being modified, fix that

File: tests/resolvers/User/post.spec.ts
Comment (AVtheking) [reply tests/resolvers/User/post.spec.ts:97]: Sir I have tried that in the first place but I don't know why the test was failing even though the result were same that's why I have tried this way

File: src/typeDefs/queries.ts
Comment (AVtheking) [reply src/typeDefs/queries.ts:110]: sorry sir but I haven;t created this ,this was already there

**Comment by AVtheking** [comment]
@xoldyckk I have made the changes please review

File: src/typeDefs/queries.ts
Comment (palisadoes) [reply src/typeDefs/queries.ts:110]: Though it's not your code, the functionality will be affected. Please fix

**Comment by palisadoes** [comment]
@AVtheking 

Please click on the resolve button on the code you have fixed. It'll be easier to do the follow up review that way.

File: src/typeDefs/queries.ts
Comment (AVtheking) [reply src/typeDefs/queries.ts:110]: so I have to change this functionality to return simple post object ?

**Comment by palisadoes** [comment]
Please fix the failing tests

**Comment by AVtheking** [comment]
image
both value are literally same and this test is not failing in my local

**Comment by palisadoes** [comment]
Is it the 4 quotes in expected string versus 2 quotes in the received string? Are objects of the same type?

**Comment by AVtheking** [comment]
For this only I tested this in my local to look into this but it passed all test cases in my local

**Comment by palisadoes** [comment]
It's probably related to differences in the two environments.

The two are not equal because of the excess quotes. You need to figure out why one uses them and the other doesn't

**Comment by AVtheking** [comment]
@palisadoes  I have fixed the failing test case

File: src/typeDefs/queries.ts
Comment (xoldyckk) [reply src/typeDefs/queries.ts:110]: @AVtheking It would break these instances in talawa-admin:-
https://github.com/search?type=code&auto_enroll=true&q=repo%3APalisadoesFoundation%2Ftalawa-admin+postsByOrganizationConnection

Comment (xoldyckk) [reply src/typeDefs/queries.ts:110]: ideally best practices and designs should be followed from the start so major disruptions aren't caused in the future, but right now talawa applications are not in that state

Right now the practice should be:-

add new functionalities without breaking the existing ones -> migrate the clients to use the new functionalities -> deprecate/remove the old functionalities

Comment (xoldyckk) [reply src/typeDefs/queries.ts:110]: > so I have to change this functionality to return simple post object ?

It's not a simple post object, it's just named incorrectly `PostConnection` instead of `PostsConnection`, like I had said `postsByOrganizationConnection` should be replaced with `Organization.posts` if possible, if you're up for the task make an issue in talawa-admin and upon merging of this PR fix the issues caused by the changes in this PR.

Comment (AVtheking) [reply src/typeDefs/queries.ts:110]: so sir  I have to change to change it to Organization.posts and fix that in admin also ?

Comment (AVtheking) [reply src/typeDefs/queries.ts:110]: image
this query is also getting the posts of the organization

Comment (xoldyckk) [reply src/typeDefs/queries.ts:110]: > so sir I have to change to change it to Organization.posts and fix that in admin also ?

yes, if you intend to remove the old field and make talawa-admin use the new field

Comment (xoldyckk) [reply src/typeDefs/queries.ts:110]: Though it's not necessary that you fix it, you could also open issues on those repositories and corresponding slack channels and inform the contributers about the breaking changes

Comment (AVtheking) [reply src/typeDefs/queries.ts:110]: image
so I have to remove that query and paste it here in the organization type as posts and move the implementation of the query inside the organization ?

Do I also need to change the pagination logic of this function ?

Comment (xoldyckk) [reply src/typeDefs/queries.ts:112]: @AVtheking just like you have done for `User.posts` you would do for `Organization.posts`, the difference would be that instead of filtering the posts by user_id(parent._id) you would be filtering with organization_id(parent._id).

Comment (AVtheking) [reply src/typeDefs/queries.ts:112]: Wouldn't it requires where as argument sir to filter the posts ?

Comment (AVtheking) [reply src/typeDefs/queries.ts]: @xoldyckk I have removed those queries and made it in the organization,If it needs to include where as argument also then let me now I will change it .

**Comment by Olatade** [comment]
@AVtheking please can you update your branch with the latest changes from develop

File: src/resolvers/Organization/posts.ts
Comment (xoldyckk) [inline src/resolvers/Organization/posts.ts:121]: get this if statement out of the previous if block, it should exist as a seperate check, do it like it is done in childTags.ts

File: src/resolvers/User/posts.ts
Comment (xoldyckk) [inline src/resolvers/User/posts.ts:121]: same as above

File: tests/resolvers/User/post.spec.ts
Comment (xoldyckk) [reply tests/resolvers/User/post.spec.ts:78]: That can't be true, the implementation is 1:1, why would it fail in the test?

File: src/typeDefs/queries.ts
Comment (xoldyckk) [reply src/typeDefs/queries.ts]: @AVtheking filtering and sorting are expensive operations and therefore should only be added when client apps need it, the fields being exposed must also be considered, not all fields are indexed in database

File: src/resolvers/Organization/posts.ts
Comment (AVtheking) [reply src/resolvers/Organization/posts.ts]: done

File: tests/resolvers/User/post.spec.ts
Comment (AVtheking) [reply tests/resolvers/User/post.spec.ts:84]: idk what's happening I even tried log the output to see any differences but data is same but still failing

**Comment by AVtheking** [comment]
I have made the changes

**Review by xoldyckk — APPROVED** [review]

**Comment by palisadoes** [comment]
There seems to be an untested error in this PR.

Please make an emergency change

**Comment by AVtheking** [comment]
yes sir done

**Pull request merged**
